app/bridge/client.js

```diff
@@ -90,85 +121,142 @@ function createNewClient(name, version) {
     this.client = client;
     this.uuid = uuid();
 
-    this.server = null;
+    this.serverChannel = null;
+
+    this.connected = false;
+    this.connectionDeferred = null;
+    this.disconnectionDeferred = null;
+
     this.connect();
   }
 
   ClientInternal.prototype.connect = function() {
-    new Registration(this.client, this.uuid);
-    this.server = new BroadcastChannel(this.uuid);
-    this.listen();
+    debug(this.client.name, this.uuid + ' [connect]');
+
+    if (this.connected) {
+      return Promise.resolve();
+    } else if (this.disconnectionDeferred) {
+      // wait for complete disconnection before trying to connect
+      return this.disconnectionDeferred.promise.then(() => this.connect());
+    } else if (this.connectionDeferred) {
+      return this.connectionDeferred.promise;
+    } else {
+      this.connectionDeferred = new Deferred();
+      register(this.client, this.uuid);
+      // It might not be the first time we connect
+      if (!this.serverChannel) {
+        this.serverChannel = new BroadcastChannel(this.uuid);
+        this.listen();
+      }
```

Comment:
With the current state, it can't because servers can disconnect while still having clients. With the logic change you suggested above (disconnect clients first then servers), yes :-)